Bitcoin (#283)
dfinity/portal: docs/developer-docs/functionality/bitcoin/t-ecdsa.md

# Threshold ECDSA

> The Internet Computer implements a novel threshold ECDSA protocol. In such protocol, the private ECDSA key is held in a secret-shared manner by multiple parties, namely the replicas of a threshold-ECDSA-enabled subnet on the IC, and signatures are computed using those secret shares without the private key ever being reconstructed. Each replica of such subnet holds a key share that provides no information about the key on its own. At least one third of the replicas are required to generate a threshold signature using their respective key shares.

Reviewer: ```suggestion
The Internet Computer implements a novel threshold ECDSA protocol. In this protocol, the private ECDSA key is held in a secret-shared manner by multiple parties, namely the replicas of a threshold-ECDSA-enabled subnet on the IC, and signatures are computed using those secret shares without the private key ever being reconstructed. Each replica of such subnet holds a key share that provides no information about the key on its own. At least one third of the replicas are required to generate a threshold signature using their respective key shares.
```

> Each canister on any subnet of the Internet Computer has control over an ECDSA key pair and can request signatures with the corresponding private key to be computed. A signature is only issued to the eligible canister, i.e., the legitimate holder of the ECDSA key. I.e., each canister can obtain signatures only for its ECDSA keys. Note that canisters do not hold any private ECDSA keys or key shares themselves, but delegate this to specific threshold-ECDSA-enabled subnets of the IC. Threshold cryptography can help enable functionality in the trust model of a blockchain that would be impossible to achieve with conventional cryptography alone.

Reviewer: ```suggestion
Each canister on any subnet of the Internet Computer has control over a unique ECDSA key pair and can request signatures with the corresponding private key to be computed. A signature is only issued to the eligible canister, i.e., the legitimate holder of the ECDSA key. Each canister can obtain signatures only for its ECDSA keys. Note that canisters do not hold any private ECDSA keys or key shares themselves, but delegate this to specific threshold-ECDSA-enabled subnets of the IC. Threshold cryptography can help enable functionality in the trust model of a blockchain that would be impossible to achieve with conventional cryptography alone.
```
AFAIU the key pairs are unique per canister, and I think it is worth pointing this out explicitly.

A threshold ECDSA implementation on a blockchain can be viewed as the on-chain pendant to a hardware security module (HSM) that stores private keys securely and issues signatures on request of the eligible entities, and only to those.

The availability of threshold ECDSA allows for a multitude of important use cases, as for example:

-   Canisters natively holding Bitcoin;

-   Integration with Ethereum, e.g., getting the ERC-20 tokens of Ethereum into the IC or signing Ethereum transactions;

> -   Realizing a decentralized certification authority (CA), where certificates are issued using threshold ECDSA (this requires a different elliptic curve to the currently implemented curve \`secp256k1\`, namely \`secp256r1\`, for use cases such as integration with Bitcoin and Ethereum).

Reviewer: ```suggestion
-   Realizing a decentralized certification authority (CA), where certificates are issued using threshold ECDSA (this requires a different elliptic curve to the currently implemented curve `secp256k1`, namely `secp256r1`, for use cases such as integration with Bitcoin and Ethereum).
```

> -   FIX: More use cases

Reviewer: just pointing out the TODO to make future reviews easier

## Technology Overview -- How It Works

> We next give a high-level outline of threshold ECDSA on the IC. Some of the information in this section is not required to use the feature, but may be of interest to the technically inclined reader for obtaining background information on the technology. The IC implements the threshold ECDSA protocol by Groth and Shoup as described in ??. The threshold ECDSA implementation on the IC features multiple protocols:

Reviewer: link to paper is still a todo

-   Signing: This protocol computes a signature with the secret-shared private key. Signing is done by precomputing the vast majority of the steps of the signing protocol and storing the precomputation result as a set of quadruples. Upon a signing request, one of the precomputed quadruples is consumed to efficiently compute a signature.

It is crucial to note that the private key never exists in reconstructed, but only in secret-shared form during its whole lifetime, be it during its generation, the re-sharing of the key from one subnet to another, and computing signatures.

Various NNS proposals have been implemented to perform key management, i.e., initial key generation and key re-sharing. Those proposals are used to define on which subnet to generate an ECDSA master key, to which subnet to re-share the key to have it available for better availability, and which subnet to enable for answering signing requests.

## ECDSA Keys

> ECDSA-enabled subnets hold what we call threshold ECDSA *master keys*, generated with the key generation process on selected subnets of the IC. A master ECDSA key is a key from which canister ECDSA keys can be derived. I.e., a single master key for a given elliptic curve suffices to allow for the derivation of an ECDSA key for each canister on the IC (*canister root key*) using the canister\'s principal and a BIP-32-like key derivation mechanism that is executed transparently by the protocol as part of the signing and public key retrieval APIs. See the level-0 key derivation in Figure ?? for the derivation of canister root keys from a master key.

Reviewer: figure number is not here yet

From a canister root key, an unlimited number of further ECDSA keys can be derived for the canister using a backward-compatible extension of the BIP-32 key derivation mechanism. The extension allows not only 32-bit integers, but arbitrary-length byte arrays, to be used as input for each level of the key derivation function. See the levels 1 and greater in the below figure illustrating the derivation of further canister keys based on the canister root key.

The derivation of further ECDSA keys from a canister root key can be done without involvement of the IC to facilitate certain use cases.

> ![](media/image1.png){width="6.5in" height="3.625in"}

Reviewer: figure file is not present

> Threshold ECDSA master keys are always referred to through *key identifiers* in the threshold ECDSA API (as well as in the NNS proposals for managing the rollout). The key identifiers comprise an elliptic curve name and an identifier, e.g., an example key identifier is the 2-tuple \`(secp256k1, key_1)\`. Those key identifiers are used by the system to refer to the correct key, e.g., for selecting the key share when computing a signature or in the implementation of the XNet routing of API calls and responses to/from the ECDSA-enabled subnet holding the key with the corresponding identifier.

Reviewer: ```suggestion
Threshold ECDSA master keys are always referred to through *key identifiers* in the threshold ECDSA API (as well as in the NNS proposals for managing the rollout). The key identifiers comprise an elliptic curve name and an identifier, e.g., an example key identifier is the 2-tuple `(secp256k1, key_1)`. Those key identifiers are used by the system to refer to the correct key, e.g., for selecting the key share when computing a signature or in the implementation of the XNet routing of API calls and responses to/from the ECDSA-enabled subnet holding the key with the corresponding identifier.
```

## Deployment

We next outline the deployment for the Chromium (Beta) release available now and the GA release coming later in 2022.

### Chromium Release

> Currently, as part of the Chromium release, only a test key for curve \`secp256k1\` is deployed on one subnet with a replication factor of 13. This key may be deleted with an according NNS proposal some time after the GA release and therefore should not be used for anything that has value, but only for development and testing purposes. More concretely, it is, for example, strongly advised against holding real bitcoin with the test key. The test key is rather intended to facilitate development of Bitcoin smart contracts and hold Testnet bitcoin as preparation for the GA release. The test key has the id \`(secp256k1, test_key_1)\` (FIX key name) for referring to it in API calls.

Reviewer: reviewer todo: check all uses of \

> Currently, as part of the Chromium release, only a test key for curve \`secp256k1\` is deployed on one subnet with a replication factor of 13. This key may be deleted with an according NNS proposal some time after the GA release and therefore should not be used for anything that has value, but only for development and testing purposes. More concretely, it is, for example, strongly advised against holding real bitcoin with the test key. The test key is rather intended to facilitate development of Bitcoin smart contracts and hold Testnet bitcoin as preparation for the GA release. The test key has the id \`(secp256k1, test_key_1)\` (FIX key name) for referring to it in API calls.

Reviewer: there's still a todo in here

### Further Aspects

For both the current Chromium deployment with a test key and the future GA release with a production key, requests to the threshold ECDSA API are always XNet requests because the ECDSA-enabled subnets do not host user\'s canisters, thus some seconds of extra latency is incurred due to Xnet communications from the calling canister\'s subnet to the threshold-ECDSA-enabled subnet holding the respective key.

> Support for further elliptic curves and additional corresponding master keys may be added in the future. The curve \`secp256r1\` is interesting for supporting use cases such as decentralized certification authorities (CAs) and is the premier candidate group to be added to facilitate use cases like the mentionedone.

Reviewer: ```suggestion
Support for further elliptic curves and additional corresponding master keys may be added in the future. The curve \`secp256r1\` is interesting for supporting use cases such as decentralized certification authorities (CAs) and is the premier candidate group to be added to facilitate use cases like the mentioned one.
```

## API

> We next give an overview of the API for threshold ECDSA. For a detailed specification, the reader is referred to the corresponding part of the [[Internet Computer Interface Specification]{.underline}](https://internetcomputer.org/docs/current/references/ic-interface-spec/#ic-ecdsa_public_key). The API comprises two methods, \`ecdsa_public_key\` for retrieving threshold ECDSA public keys, and \`create_ecdsa_signature\` for requesting threshold ECDSA signatures to be computed from the subnet holding the secret-shared private threshold ECDSA key.

Reviewer: ```suggestion
We next give an overview of the API for threshold ECDSA. For a detailed specification, the reader is referred to the corresponding part of the [Internet Computer Interface Specification](../../../references/ic-interface-spec.md#ic-ecdsa_public_key). The API comprises two methods, \`ecdsa_public_key\` for retrieving threshold ECDSA public keys, and \`create_ecdsa_signature\` for requesting threshold ECDSA signatures to be computed from the subnet holding the secret-shared private threshold ECDSA key.
```

    > This call requires that the ECDSA feature is enabled, and the \`canister_id\` meets the requirement of a canister id. Otherwise it will be rejected.

-   \`sign_with_ecdsa\`: This method returns a new ECDSA signature of the given message_hash that can be separately verified against a derived ECDSA public key. This public key can be obtained by calling \`ecdsa_public_key\` with the caller\'s \`canister_id\`, and the same \`derivation_path\` and \`key_id\` used here.\
    > The signatures are encoded as the concatenation of the SEC1 encodings of the two values r and s. For curve \`secp256k1\`, this corresponds to 32-byte big-endian encoding.\
    > This call requires that the ECDSA feature is enabled, the caller is a canister, and \`message_hash\` is 32 bytes long. Otherwise it will be rejected.

## Environments

In order to facilitate developers throughout the canister development life cycle on the IC, the feature is available in both the SDK for local development and testing as well as on IC mainnet for pre-production testing and production operation of canisters.

### SDK

> The development of canisters is typically done in the local environment of the developer, facilitated by use of the SDK. The SDK has been extended such that the management canister API for threshold ECDSA is available in the local canister execution environment. Thus, canisters using the threshold ECDSA API can be run locally for development and testing purposes. The feature is always enabled in the SDK, so no further user action is required in order to make use of the API.

Reviewer: ```suggestion
The development of canisters is typically done in the local environment of the developer, facilitated by use of the SDK. The SDK has been extended such that the management canister API for threshold ECDSA is available in the local canister execution environment. Thus, canisters using the threshold ECDSA API can be run locally for development and testing purposes. To enable the feature, look at the output of `dfx start --help` for instructions.
```

### IC Mainnet

Any canister on any subnet of IC mainnet can call the threshold ECDSA API exposed by the management canister. The calls are routed via XNet communication to the ECDSA-enabled subnet that holds the key referred to in the API call (only one such signing subnet holding a test key exists for testing purposes in the initial Chromium release). Note that this test key is hosted on a subnet with a replication factor of only 13 and may be deleted in the future, thus it should not be used for anything of value, but rather solely for development and testing purposes. The main intended purpose is to facilitate the development and testing of Bitcoin-enabled dApps using Bitcoin testnet.

Later in 2022, as part of the General Availability (GA) release of the feature, a production ECDSA key on the \`secp256k1\` elliptic curve will be deployed to be used for integration with Bitcoin mainnet and other use cases of interest.

> ## Code Walkthrough

Reviewer: assuming the rest will be fleshed out later

### Getting Started

Sample project called `threshold-ecdsa` is provided in the [examples repository](??), under either `motoko` or `rust` sub-directories. It requires at least dfx version ?? for local development.

> -   Download and install SDK if not already have.
> -   Check out the examples repistory.
> -   Change directory to `motoko/threshold_ecdsa` subdirectory.
> -   Run `dfx deploy`.

Reviewer: ```suggestion
Please refer to the instructions in the example's README.md for instructions how to run the sample code.
```

### Deploying it on Mainnet

> The same code used in the SDK environment for local testing works IC on mainnet. Keep in mind that the master key in the SDK environment is different to the master test key on mainnet and the future master production key.

Reviewer: ```suggestion
The same code used in the SDK environment for local testing works on IC mainnet. Keep in mind that the master key in the SDK environment is different to the master test key on mainnet and the future master production key.
```

> Before deploying to mainnet, you should modify the code to use right name of the `key_id` to use.

Reviewer: ```suggestion
Before deploying to mainnet, you should modify the code to use right name of the `key_id`.
```
maybe also add what the right key_id is or how to figure it out?
